pkg/controller/app_controller.go

```diff
@@ -0,0 +1,221 @@
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/klog"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/juicedata/juicefs-csi-driver/pkg/config"
+	"github.com/juicedata/juicefs-csi-driver/pkg/k8sclient"
+	"github.com/juicedata/juicefs-csi-driver/pkg/util"
+)
+
+type AppController struct {
+	*k8sclient.K8sClient
+}
+
+func NewAppController(client *k8sclient.K8sClient) *AppController {
+	return &AppController{client}
+}
+
+func (a *AppController) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	klog.V(6).Infof("Receive pod %s %s", request.Name, request.Namespace)
+	pod, err := a.K8sClient.GetPod(ctx, request.Name, request.Namespace)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		klog.Errorf("get pod %s error: %v", request.Name, err)
+		return reconcile.Result{}, err
+	}
+	if pod == nil {
+		klog.V(6).Infof("pod %s not found.", request.Name)
+		return reconcile.Result{}, nil
+	}
+
+	if !ShouldInQueue(pod) {
+		klog.V(6).Infof("pod %s namespace %s should not in queue", request.Name, request.Namespace)
+		return reconcile.Result{}, nil
+	}
+	// umount fuse sidecars
+	err = a.umountFuseSidecars(pod)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	return reconcile.Result{}, nil
+}
+
+func (a *AppController) umountFuseSidecars(pod *corev1.Pod) (err error) {
+	for _, cn := range pod.Spec.Containers {
+		if strings.Contains(cn.Name, config.MountContainerName) {
+			if e := a.umountFuseSidecar(pod, cn); e != nil {
+				return e
+			}
+		}
+	}
+	return
+}
+
+func (a *AppController) umountFuseSidecar(pod *corev1.Pod, fuseContainer corev1.Container) (err error) {
+	if fuseContainer.Name == "" {
+		return
+	}
+
+	cmd := []string{}
+	// get prestop
+	if fuseContainer.Lifecycle != nil && fuseContainer.Lifecycle.PreStop != nil && fuseContainer.Lifecycle.PreStop.Exec != nil {
+		cmd = fuseContainer.Lifecycle.PreStop.Exec.Command
+	}
+
+	klog.Infof("[AppController] exec cmd [%s] in container %s of pod [%s] in [%s]", cmd, config.MountContainerName, pod.Name, pod.Namespace)
+	stdout, stderr, err := a.K8sClient.ExecuteInContainer(pod.Name, pod.Namespace, fuseContainer.Name, cmd)
+	if err != nil {
+		if strings.Contains(stderr, "not mounted") ||
+			strings.Contains(stderr, "mountpoint not found") ||
+			strings.Contains(stderr, "no mount point specified") {
+			// if mount point not mounted, do not retry
+			return nil
+		}
+		if strings.Contains(err.Error(), "exit code 137") {
+			klog.Warningf("[AppController] exec with exit code 137, ignore it. error: %v", err)
+			return nil
+		}
+		klog.Errorf("[AppController] error: %v; exec stdout: %s; exec stderr: %s", err, stdout, stderr)
+		return err
```

Comment:
Should removed as duplicated with the line 109

Reply:
done